RFC: Basic NOBIL implementation

robho opened the pull request:
This adds a [NOBIL](https://info.nobil.no/english) data source. Data is available in Denmark, Finland, Iceland, Norway and Sweden and the data is collected both from operators via OCPI and contributed by the public.

This is a draft implementation where some functionality is missing. I'm interested in hearing if this is of interest, and if so, what should I fix/work on going forward?

Known problems / missing features:
* Not all charger data is exposed in the app (photo, location (near restaurant/shopping/...), payment methods, ...)
* No chargers visible in Iceland although the data exists
* No filters are implemented
* Chargeprice connection missing
* Does Android Auto work? Not yet tested
* The code could use some cleanup from someone with Android development experience

johan12345 commented:
Very nice, thanks a lot! I'll try to find some time in the weekend to try it out and provide some help with code cleanup.

> Not all charger data is exposed in the app (photo, location (near restaurant/shopping/...), payment methods, ...)

Yeah, if there are properties that don't fit into EVMap's data structure we can think about adding new fields or expanding existing ones where it is useful. For payment methods, we have the `chargecards` structure (used by GoingElectric, but a bit more complex than necessary for what NOBIL provides) or we could just put it into the `cost.description` for now). For location, the most similar field in GoingElectric would be "category", but that is not exposed through their API, so I haven't added it so far. `amenities` might also work, though that is usually a slightly longer text description in GoingElectric.

> No filters are implemented

As previously mentioned, I guess all the filters would have to be implemented locally, as the Nobil API doesn't really have server-side filtering abilities, right? That should be possible (just like the min_connectors filter for GE and OCM), but in the longer term the better solution is probably to first download all chargers into the local DB (which is a capability I'm implementing for #290) and perform filtering there.

> Chargeprice connection missing

Chargeprice currently only support GoingElectric and OpenChargeMap unfortunately. And they by now they have also created their own separate database of charging stations aggregated from different sources, so from what I heard they are not so interested in fixing issues with their adapters for GE and OCM, let alone adding new ones 🫤.
Also, there is #320 - so I am not sure how long the native Chargeprice integration in EVMap will stay or whether I will have to try to build an alternative at some point.

> Does Android Auto work? Not yet tested

Likely yes, I can test that as well.

> When a charging point is removed in nobil it doesn't disappear in the UI (a cache clear is needed). Maybe a generic evmap bug?

True, that is probably a general bug - for GoingElectric it's not so relevant as we are only allowed to cache the data for 24h anyway. But yeah, it probably also applies to OpenChargeMap. Let's track that in a separate issue.

johan12345 commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilApi.kt:
```diff
@@ -0,0 +1,215 @@
+                    val new = original.newBuilder()
+                        .build()
+                    chain.proceed(new)
+                }
```
This interceptor is not necessary here - in other APIs it is used to automatically add the API key to every request, but here the API key is part of the request body so that's not really possible.

johan12345 commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilModel.kt:
```diff
@@ -0,0 +1,234 @@
+        if (chargerStationAttributes.st["24"]?.attrTrans == "Yes") OpeningHours(twentyfourSeven = true, null, null) else null,
+        // 7: Parking fee
+        if (chargerStationAttributes.st["7"]?.attrTrans == "No") Cost(freeparking = true) else null,
+        null,
```
We have to state the data license (returned from the API under `Rights`)

robho commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilApi.kt:
```diff
@@ -0,0 +1,215 @@
+        referenceData: ReferenceData,
+        id: Long
+    ): Resource<ChargeLocation> {
+        // TODO: Nobil ids are "SWE_1234", not Long
```
I dumped data from Finland, Norway and Sweden and checked the `id` of all chargers -> the `id` field is unique across countries. I don't understand the need for two ids either..

I also verified that sending this get-charger-details-request really does require the `International_id`. No data is returned if the `id` is sent in the request.

One question I have here is whether this function is used/needed for nobil. All data that nobil returns in responses is "detailed" so is there a need to implement this function?

Thanks for your input and review! I'll go through it all, but could take some time, depending on the time I have to look into it.

robho commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilModel.kt:
```diff
@@ -0,0 +1,234 @@
+                listOfNotNull(chargerStationData.street, chargerStationData.houseNumber).joinToString(" ")),
+        chargerStationAttributes.conn.map { createChargepointFromNobilConnection(it.value) },
+        null,
+        "",
```
There's a web page to report errors for Swedish chargers (https://www.energimyndigheten.se/klimat/transporter/laddinfrastruktur/registrera-din-laddstation/elbilsagare/), but for nobil in general they want error reports by email.

Do you think it adds any value to add the web page link for Swedish chargers? Can something be done to send an email? Create a mailto link?

johan12345 replied:
Yeah, we could put this link as the `editUrl` for chargers in Sweden and use a mailto URL for the other countries. I'm not sure if Android will handle a mailto URL out of the box, but if not, we could [parse it](https://developer.android.com/reference/android/net/MailTo) in the app and convert it into an [email intent](https://developer.android.com/guide/components/intents-common#Email).

robho replied:
I made an attempt to implement the mailto handling. Comments are welcome.

johan12345 replied:
looks good!

robho replied:
I added Chargelocation.dataSourceUrl as suggested in https://github.com/ev-map/EVMap/pull/363/commits/5cc8322f37a15cf27a627b22e8c9bf85c041c90f. Does it match what you were thinking of?

By "disabling" the share button, should I do something else than just skipping "openurl"?

(I could merge the db revisions if we want to squash all nobil commits later)

johan12345 replied:
Yes, that looks great, thanks!

Hmm, I guess we have three options for the share button, not sure which would be the most intuitive:
- If `charger.url == null` and the share button is pressed, show a Snackbar/Toast, stating that this data source does not support sharing
- If `charger.url == null`, mark the share button as "disabled":
```kotlin
binding.detailAppBar.toolbar.menu.findItem(R.id.menu_share).isEnabled = false
```
- If `charger.url == null`, hide the share button entirely:
```kotlin
binding.detailAppBar.toolbar.menu.findItem(R.id.menu_share).isVisible = false
```
What do you think?

robho replied:
I think I would prefer to remove the icon. If it's disabled I would wonder what's wrong and if there's a way to enable it. A toast could be nice and the UI is then consistent across all data source implementations, but I think it gets a little confusing with UI elements that can't be used.

But I don't have a strong opinion :-)

robho commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilModel.kt:
```diff
@@ -0,0 +1,234 @@
+                "50" -> Chargepoint.SCHUKO // Type 2 + Schuko
+                "60" -> Chargepoint.CCS_UNKNOWN // Type1/Type2
+                "70" -> "" // Hydrogen
+                "82" -> "" // Biogas
```
Sounds good. There's also attribute 26, "Energy carrier", that could be interesting to look at here.

Related question.. The icon for TYPE_2_SOCKET and TYPE_2_PLUG seems to be the same. Is that intentional? I found a charging station that had both types and then the UI looked confusing.

robho replied:
I implemented you suggestion, ignoring hydrogen/biogas chargepoints.

(attribute 26, "Energy carrier" is not reliable (it's wrong for at least one charge location) so I don't look at this value)

johan12345 replied:
> Related question.. The icon for TYPE_2_SOCKET and TYPE_2_PLUG seems to be the same. Is that intentional? I found a charging station that had both types and then the UI looked confusing.

Right, good question. Yeah, it's probably good to add a different icon for TYPE_2_PLUG, though I'm not quite sure how it should look to still fit in with the style of the others... Maybe just the Type 2 icon with colors inverted? Or something with a small extra icon in the corner (like in the Tesla Supercharger icon)?

robho commented:
Regarding filtering..
> in the longer term the better solution is probably to first download all chargers into the local DB (which is a capability I'm implementing for https://github.com/ev-map/EVMap/pull/290) and perform filtering there.
Is your idea to replace the on-demand-loading with a full data fetch or are you thinking of having some combination of the two?

The nobil data is ~130 Mbytes today (up ~6 Mbytes in ~2 months). This is uncompressed data, I haven't checked if the server can compresss data and what the compressed size is. I think I would prefer to use on-demand loading most of the time, but it would be valuable to have the possibility to cache all data when needed (no mobile connection or high mobile roaming prices, ...).

Is the problem that you see with local filters that they need to fetch data for chargers that are uninteresting? Ie unnecessary network traffic and data processing? Are those "unnecessary" downloads cached even if they are not used?

robho commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilModel.kt:
```diff
@@ -0,0 +1,234 @@
+                chargerStationData.zipCode,
+                listOfNotNull(chargerStationData.street, chargerStationData.houseNumber).joinToString(" ")),
+        chargerStationAttributes.conn.map { createChargepointFromNobilConnection(it.value) },
+        null,
```
I extracted the `network` information available in nobil and there are ~1500 different "networks". The field is a text field and users have entered information in many different variations (different capitalization, typos, abbreviations, alternative names, "owned by" information, ...).

Even if I only parse operator from OCPI data I get ~60 "networks" and also here there are name variations for the same actual network and some names don't even correspond to a network.

Maybe we could create a network list manually and try to match charge locations to the list items. Not a great solution since it requires manual work to update the network list and it's hard to know which networks to include in the list, especially in "foreign" countries.

Maybe we have to live without a network filter, for now

johan12345 replied:
Ah 😟 Wow, strange that that happens even with OCPI data...
Yeah, then we probably have to live without a network filter for now, and hope that Nobil can improve the data quality in the future.

robho commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilModel.kt:
```diff
@@ -0,0 +1,234 @@
+        // 24: Open 24h
+        if (chargerStationAttributes.st["24"]?.attrTrans == "Yes") OpeningHours(twentyfourSeven = true, null, null) else null,
+        // 7: Parking fee
+        if (chargerStationAttributes.st["7"]?.attrTrans == "No") Cost(freeparking = true) else null,
```
Looks like I **either** add `freecharging` and `freeparking` **or** descriptions. If I set `freecharging` and `freeparking` then EVMap creates a text string for me and ignores the descriptions. Right?

robho replied:
I added some information to the Cost description:
> Accepted payment methods: Bank card, Charging card

I looked into "Time limit" and "Accessibility", but they were not reliable and consistent.

johan12345 commented:
> The nobil data is ~130 Mbytes today (up ~6 Mbytes in ~2 months). This is uncompressed data, I haven't checked if the server can compresss data and what the compressed size is.

Gzip usually helps quite a lot with JSON data, and the Nobil server does seem to support it. According to
```
curl --compressed -so /dev/null "https://www.nobil.no/api/server/datadump.php?apikey=<key>&fromdate=2005-01-01&format=json" -w '%{size_download}'
```
the download size is pretty small at just ~5 MB. Of course it might occupy a bit more space in the local database (and from my experience with the implementation for OSM, parsing the JSON and inserting chargers into the DB also takes much more time than the download itself).

> Is your idea to replace the on-demand-loading with a full data fetch or are you thinking of having some combination of the two?

In general, EVMap will continue to support both. So we could offer both options for Nobil. On the other hand, with such a small download size, there's not much of a downside to storing it all locally - might also reduce the load on Nobil's servers in the long term.

> Is the problem that you see with local filters that they need to fetch data for chargers that are uninteresting? Ie unnecessary network traffic and data processing?

Yeah. I think the data processing (JSON parsing into Kotlin objects, then filtering) is usually the main bottleneck when loading a large map region with thousands of chargers - server-side filtering and even server-side clustering makes it a lot faster for GoingElectric where that is supported.

Also, depending on what filters we implement, there might be some filters where we can only determine the available options by iterating through the whole dataset once (e.g., the network filter mentioned in the comments) and then storing that as `ReferenceData`.

> Are those "unnecessary" downloads cached even if they are not used?

not at the moment, because local filtering happens inside the API implementation

robho commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilModel.kt:
```diff
@@ -0,0 +1,234 @@
+                else -> null
+            }
+
+            return Chargepoint(connectionType, connectionPower, 1, null, null)
```
I added parsing of voltage/current. Then I added parsing of EVSEID into the Chargepoint structure. Does it make sense to add it as a list in Chargepoint?

johan12345 replied:
Yes, makes sense! I guess with other data sources we might also have the case where a list of EVSEIDs is given, but the mapping which EVSEID belongs to which type of plug/socket is missing - then we'll have to handle that differently. But if we have the mapping available, it's definitely good to use it.

johan12345 commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilModel.kt:
```diff
@@ -111,7 +112,10 @@ data class NobilChargerStation(
-            null,
+            when (chargerStationData.landCode) {
+                "SWE" -> "https://www.energimyndigheten.se/klimat/transporter/laddinfrastruktur/registrera-din-laddstation/elbilsagare/"
+                else -> "mailto:post@nobil.no?subject=" + URLEncoder.encode("Regarding charging station " + chargerStationData.internationalId, "UTF-8").replace("+", "%20")
```
Is the `.replace("+", "%20")` necessary? `Uri.encode` might also work as an alternative

robho replied:
Uri.encode works better. I switched.

johan12345 commented on app/src/main/java/net/vonforst/evmap/fragment/MapFragment.kt:
```diff
@@ -1001,6 +1001,8 @@ class MapFragment : Fragment(), OnMapReadyCallback, MenuProvider {
             onChargerClick = {
                 vm.chargerSparse.value = it
+                binding.detailAppBar.toolbar.menu.findItem(R.id.menu_share).isVisible =
+                    it.url != null
```
Let's rather do this in the observer (`vm.chargerSparse.observe` call in `setupObservers`). That is where the state of the favorite button in the toolbar is updated as well.

robho commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilModel.kt:
```diff
@@ -0,0 +1,234 @@
+        null,
+        Instant.now(),
+        true
+    )
```
See https://github.com/ev-map/EVMap/pull/363/commits/70953af3b2aefe4cf0ee14c8f7f9cdfab59f267e. What do you think?

I kept the name "availability" although it could perhaps be confusing when the realtime data code refers to "availability" also.

johan12345 replied:
Looks good! Regarding the naming... Hm - maybe something like `access` instead of `availability`? i.e., a charger could be "available" (not in use) but not publicly "accessible".

robho replied:
There's an "Accessibility" field in nobil that is used to specify how to to start the charging (values: Open, Standard key, Other, RFID, Payment, Cellular phone).

Currently I don't parse the nobil "Accessibility" field so there's no conflict right now.

I guess the important part is to keep EVMap code consistent with its own nomenclature. I'll take a look at renaming availability -> access.

robho replied:
I changed `availability` -> `accessibility`. Hope that works..

robho commented:
I rebased to latest master.

Some comments on this implementation and nobil..

* In Sweden the quality of nobil has increased quite much over the past 6 months. Lots of duplicates and invalid chargers have been removed and new OCPI chargers are added daily. Today there are 5313 "OCPI chargers" and 1475 manually added chargers.
* Realtime data from enbw seems to work well with nobil and coverage is good. Getting realtime from nobil would add some additional realtime data though.
* The speed of the nobil implementation isn't that great (I've only run debug builds with LeakCanary active though).
  * [The web version](https://www.nobil.no/api/client/search_apiVer3.php) is faster. For every server request that the web version sends it tells the server which chargers it already knows about (url parameter excludeIds) -> less network traffic and less data to parse.
  * EVMap caches query results, but only if the result is complete. Nobil can at most return 2000 results in a query response. If we're at a zoom level where we hit that limit (there are ~2000 chargers in Stockholm area) results aren't cached and there will be lots of queries as the user browses the map.
  
  Would it make sense to look into adding support for excludeIds url parameter? Or, would it be better to change the implementation to download a full nobil dump and work with that rather than doing on-demand server queries? (I started looking into adding support for excludeIds url parameter, but it got a bit complicated with the merging of database and network chargelocations in EVMap so I never finished it, but I may return to look at it if it would be of value..)

johan12345 commented:
Hi, sorry for the late reply!
- I guess the full download option is a good idea and maybe easier to implement than the `excludeIds` parameter. As mentioned, I am building the same for the OpenStreetMap data source in #290 - in principle it works, the main thing I'm still working on (when I have time) is to speed up marker clustering at low zoom levels by using Spatialite queries for that instead of loading everything into memory first.
- Regarding the Nobil realtime data, I started building a backend based on your prototype webserver, haven't yet deployed it to a server though. In the longterm I might want to extend it to aggregate data from different sources (both static information and realtime data) - especially now that more and more CPOs are starting to provide open data as EU regulations require them to do so.

robho commented on app/src/main/java/net/vonforst/evmap/api/nobil/NobilApi.kt:
```diff
@@ -0,0 +1,332 @@
+    @GET("datadump.php")
+    suspend fun getAllChargingStations(
+        @Query("apikey") apikey: String,
+        @Query("countrycode") countrycode: String = "DAN",
```
This line should be removed. It's only there to bypass a NetworkOnMainThreadException I can't figure out.

robho commented:
> I just added a missing new column (`coordinatesProjected`) in DB migration 26, and simplified migration 27 to simply use `ALTER TABLE` instead of recreating the table. Also, I fixed the CI build (dummy entry for the Nobil API key was missing in `_ci/apikeys_ci.xml`).

Thanks! :+1: 

> There still seems to be some issue with the DB migration ...

Oh! I hadn't seen this since I haven't been running the db migration code for a long time, but I just tried the steps you provided and I see the problem. I'll see if I can figure something out, but don't hold your breath :-)

johan12345 commented:
Ah, found it: The issue was probably a combination of what you mentioned (using `DropGeoTable()` to drop the old SpatialIndex correctly) and also having to explicitly specify `AND f_geometry_column = 'coordinates'` in any SpatialIndex queries to make sure we are querying the correct column (seems like after the migration it was querying the `coordinatesProjected` column).

johan12345 commented:
...and I also added German translations and implemented a solution to remove deleted chargers from the DB, at least for when `fullDownload` is used.

It seems to be working very well now, at least as a first MVP 🙂. @robho Do you think anything else is missing before this can be merged?

robho commented:
Great Johan! Thanks for taking a look at and fixing the migration problem! :+1:

Yes, I think this is in a good enough shape to be merged. There are a few things that I wanted to take a closer look at or get your opinion on, but that could be handled later, if you're fine with that. It's my intention to continue improving the nobil implementation, although I don't have much time to spare.

The TODOs on my list are:
* The Tesla availability code for nobil isn't great, it has lots of duplication. I wanted to see if it could be merged in some way with the generic code (now there's a `if (nobil) {` branch). I also noticed that Tesla chargers have different connector configuration in nobil for different countries and wanted to look into this and work with the nobil people to get rid of the differences, if that's ok with them. The inconsistent connector handling for Tesla chargers in nobil also means that not all Tesla stations get real-time data (I guess, I haven't checked).
* I added payment methods to the `Cost` description (your suggestion a long time ago), but it's all in English right now. I started looking into translating the strings but couldn't translate strings in NobilModel (a design decision I think), instead it looks like translations are in the `Cost` object itself. I guess I could push the payment methods to the `Cost` object and have it build the description, but I got the feeling that it would get a bit messy (but I never finished it).
* The `verified` trait for nobil chargers doesn't match the documentation (FAQ) since I set verified if the charger is received through OCPI or has been manually updated recently (6 months). My idea was that this would in some way signal that the charger data is reliable.

Any comments on the bullets above would be appreciated.
